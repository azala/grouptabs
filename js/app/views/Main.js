define([
	"ring",
	// "dojo/_base/declare",
	// "dojo/dom-construct",
	// "dojo/number",
	"./_Scene"
	// "dojo/text!./templates/Main.html"
], function(ring/*declare, domConstruct, number*/, _Scene/*, template*/){

return ring.create([_Scene], {

	// templateString: template,

	name: "main",
	
	postCreate: function(){
		this.$super()
		this.summaryNode = this.domNode
		this._renderSummary()
	},
	
	_onChangeTabClick: function(){
		this.close(this, "tabs")
	},
	
	_onNewEntryClick: function(){
		this.close(this, "editEntry")
	},
	
	_onSummaryClick: function(){
		this.close(this, "summary")
	},
	
	_onListClick: function(){
		this.close(this, "list")
	},
	
	onShow: function(){
		this.app.setHomeView("main")
	},

	refresh: function(){
		// this.tabNameNode.innerHTML = this.app.tab
		this.refreshSummary()
	},
	
	refreshSummary: function(){
		this._renderSummary()
	},

	_setEmpty: function(isEmpty){
		// if(isEmpty){
		// 	this.contentNode.style.display = 'none'
		// 	this.emptyNode.style.display = ''
		// }else{
		// 	this.contentNode.style.display = ''
		// 	this.emptyNode.style.display = 'none'
		// }
	},

	_renderSummary: function(){
		this.app.getAccounts()
		.then(function(accounts){
			this._clearSummary()
			this._setEmpty(!accounts.length)
			var maxAmount = (function(){
				var result = 0
				accounts.forEach(function(account){
					var amount = Math.abs(account.amount)
					if(amount > result){
						result = amount
					}
				})
				return Math.round(result * 100) / 100
			})()
			var cssColor = function(amount){
				if(!maxAmount){
					return 'transparent'
				}
				var isNegative = amount < 0
				amount = Math.abs(amount)
				var factor = amount / maxAmount
				var hue = isNegative ? 19 : 98  // red: 360, green: 120
				var saturation = isNegative ? '77%' : '88%'
				var levelFactor = isNegative ? 0.99 : 0.93
				var level = Math.round(100 - factor * 50 * (1 - levelFactor + 1)) + '%'  // 50% is full color
				return "hsl(" + hue + ", " + saturation + ", " + level + ")"
			}
			var html = "<table id='balance'>"
			accounts.forEach(function(accountObj){
				var amount = /*number.format(*/accountObj.amount/*, {places: 2})*/
				html +=
					"<tr style='background-color: " + cssColor(accountObj.amount) + "'>"
					+ "<th class='account'>" + accountObj.participant + "</th>"
					+ "<td class='amount'>" + amount + "</td>"
					+ "</tr>"
			})
			html += "</table>"
			this.summaryNode.innerHTML = html
		}.bind(this))
	},
	
	_clearSummary: function(){
		this.summaryNode.innerHTML = ""
	}
})

})
