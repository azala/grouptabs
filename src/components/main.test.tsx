import React from "react";
import renderer from "react-test-renderer";
import Main from "./main";

it("renders empty view with no tab selected", () => {
  const tree = renderer
    .create(
      <Main
        accounts={[]}
        transactions={[]}
        total={0}
        onChangeTabClick={jest.fn()}
        onNavigateToAddTransaction={jest.fn()}
        onDetailsClick={jest.fn()}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders missing tab info error", () => {
  const tree = renderer
    .create(
      <Main
        tabId="1"
        accounts={[]}
        transactions={[]}
        total={0}
        onChangeTabClick={jest.fn()}
        onNavigateToAddTransaction={jest.fn()}
        onDetailsClick={jest.fn()}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
