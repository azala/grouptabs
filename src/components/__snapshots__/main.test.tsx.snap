// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders empty view with no tab selected 1`] = `
<div
  className="scene mainScene hidden"
>
  <div
    className="header"
  >
    <button
      className="left"
      onClick={[MockFunction]}
    >
      <svg
        height={16}
        width={16}
      >
        <path
          d="m2 2c-0.554 0-1 0.446-1 1s0.446 1 1 1h12c0.554 0 1-0.446 1-1s-0.446-1-1-1h-12zm0 5c-0.554 0-1 0.446-1 1s0.446 1 1 1h12c0.554 0 1-0.446 1-1s-0.446-1-1-1h-12zm0 5c-0.554 0-1 0.446-1 1s0.446 1 1 1h12c0.554 0 1-0.446 1-1s-0.446-1-1-1h-12z"
        />
      </svg>
    </button>
    <h2>
      
    </h2>
    <button
      className="create"
      onClick={[Function]}
    >
      +
    </button>
  </div>
  <div>
    <div
      className="load-error"
    >
      <img
        alt=""
        src="logo.png"
      />
      <h2>
        Grouptabs
      </h2>
      <p>
        Error: Tab data missing. Are you offline? Try refreshing.
      </p>
      <button
        className="create"
        onClick={[Function]}
      >
        OK
      </button>
    </div>
  </div>
</div>
`;

exports[`renders missing tab info error 1`] = `
<div
  className="scene mainScene hidden"
>
  <div
    className="header"
  >
    <button
      className="left"
      onClick={[MockFunction]}
    >
      <svg
        height={16}
        width={16}
      >
        <path
          d="m2 2c-0.554 0-1 0.446-1 1s0.446 1 1 1h12c0.554 0 1-0.446 1-1s-0.446-1-1-1h-12zm0 5c-0.554 0-1 0.446-1 1s0.446 1 1 1h12c0.554 0 1-0.446 1-1s-0.446-1-1-1h-12zm0 5c-0.554 0-1 0.446-1 1s0.446 1 1 1h12c0.554 0 1-0.446 1-1s-0.446-1-1-1h-12z"
        />
      </svg>
    </button>
    <h2>
      
    </h2>
    <button
      className="create"
      onClick={[Function]}
    >
      +
    </button>
  </div>
  <div>
    <div
      className="load-error"
    >
      <img
        alt=""
        src="logo.png"
      />
      <h2>
        Grouptabs
      </h2>
      <p>
        Error: Tab data missing. Are you offline? Try refreshing.
      </p>
      <button
        className="create"
        onClick={[Function]}
      >
        OK
      </button>
    </div>
  </div>
</div>
`;
